### ISSUE TITLE
Particle size_var isn't used.

### ISSUE THREAD

**Issue by caffeinepills** [los-cocos/cocos#259]
Looking at the code there is a size_var to cause a variation in size. However, this doesn't actually affect anything. Looking at the draw command, cocos just forces all points to a certain size through `glPointSize(self.get_scaled_particle_size())`

**Comment by dangillet** [comment]
Hello,

Good point (pun intended ;) ).

Looking at the code, it seems that making variable sized points is not easy, as they're using `glDrawArrays`  which does not support changing the point size in an array, AFAIK.

Possible solution would be to write shaders instead. But that's quite an endeavour. Not something I could do, unfortunately.

**Comment by dangillet** [comment]
Hello,

I took the challenge to try to solve this problem. I think I could come up with a pretty good solution. I'm pushing it in a *particles* branch, waiting for Claudio to make his comments. But in the meantime, you could pull that branch and see if it works for you.

As in the original code, there are 2 methods : one preferred and one fall-back. The preferred method now uses shaders which are compatible with OpenGL 2.1. The fall-back method has been improved also to better use Numpy, pushing some python for loops in the C code.

Dan.

**Comment by caffeinepills** [comment]
Shaders do sound like a solution, although I don't have experience in that area. Thanks for looking into this.

Anyways, I've tested your particle branch and although it works, it sets all of the particles at a set size of 1 pixel. (self.particle_size_idx) (Testing through the test/test_particle_* files)

**Comment by dangillet** [comment]
Hello,

Thanks for testing !

I'm not sure I'm following you 100%. `self.particle_size_idx` is actually the index (`GLuint`) that openGL returns to allow us to pass data from the program to the shader. In our case we need it to pass the `self.particle_size` array to the shader. `self.particle_size_idx` value has nothing to do with particles size. The particles sizes are set as initially designed : by setting `Particle.size` and `Particle.size_var`.

If in the examples you would change `size_var` for one of the `ParticleSystem`, you would clearly see that particles have different sizes.

**Comment by caffeinepills** [comment]
Sorry, let me clarify. I pulled down your branch and used the existing test files. For example, Spiral, which has 
```
    size = 20.0
    size_var = 10.0
```
Sets all to the size of 1 pixel. 

Even if I change the size_var to 0, It sets it to 1 as well. This should be straight forward to reproduce by just running those tests. Incase you are getting different results, here is a picture: http://imgur.com/8255EEG

**Comment by dangillet** [comment]
Hello,

Thanks for the explanation. Indeed, on my computer I don't have the same result. Here is a screenshot of what I get when running `test_particle_spiral.py`. https://www.dropbox.com/s/u38edegszw2tsup/cocos_test_test_particle_spiral.py.png?dl=0

If you're willing to help finding what the problem might be, could you let me know what OS and graphic card you have ? I assume you have at least OpenGL 2.1. Today we are at version 4.5, so I guess this is a safe assumption.

For some troubleshooting, could edit `particle.py` and change on line 56 `forced_point_sprites = False` ? This will force the particles to be rendered using quads instead of using the shader. This could already help me trying to figure out what gets wrong on your side.

**Comment by caffeinepills** [comment]
I've tried it on two different computers, although they both have NVIDIA graphics cards. Anyhow, on my home PC it is Windows 7 Professional 64 bit.  NVIDIA Geforce GTX 670 (latest video drivers) . Indeed if I change forced_point_sprites to false, it works correctly.

**Comment by dangillet** [comment]
Thank you very much for taking the time to test this. I've made a change in the code, enabling point size to be defined in shaders.

Luckily I could reproduce your problem on my other computer with a Nvidia graphic card. With this change, it seems to be resolved on my side. Please let met know how it goes on your side.

**Comment by caffeinepills** [comment]
That resolved the issue for me as well! Size_var now works as well. Thanks for taking the time to look into this!

**Comment by ccanepa** [comment]
reopened to see how to integrate in cocos ( caffeinepills : if you don't want to receive mails from this issue, go to the github page for the issue and click the 'unsubscribe button' (and thanks for your report and testing))

Daniel : I will look at the code in this weekend, promise

**Comment by dangillet** [comment]
Thanks Claudio. I could be interesting to talk a bit about what could be done in cocos. What would be the best way to get in touch ? Email ?

**Comment by ccanepa** [comment]
Yes, email would be the best

**Comment by dangillet** [comment]
I just don't find the way to send you a private message, and I don't have your email. So here is mine: dan dot gillet737 at gmail dot com. I'll delete this message once you have it. :)

**Comment by ccanepa** [comment]
__performance__

looks good
```
                            fps
	 			 point_spr		   quads
				new		old		new		old
	fireworks	350		360		270		63
	fire		270		284		268		212
	flower		350		403		365		215
	smoke		440		480		460		390
	spiral		410		440		370		222
	sun			390		404		375		250
	meteor 		405		444		412		320
	galaxy 		405		455		415		320
```

will continue later, must go

**Comment by dangillet** [comment]
Thanks for these tests ! I must say I felt very proud of the results for the new quads, but I have slightly disappointed by the new point sprites. But then I realized that the new code is doing **more** than the previous one. It has to make point sprite calculation for each individual particles, instead of just taking one number for all of them.

I realized that there is one minor improvement I still should do : the current code is not using anymore `get_scaled_particle_size`, which means that if we would zoom on the particles, their point size would not change. I could try to add this back in the code. It shouldn't be too complicated I think.

**Comment by dangillet** [comment]
Hello Claudio,

As promissed I've added support for particle resize in respect to node scaling and window resize. As in the former code, particle resize is only supported for the *main* drawing method (now with shader) and not for the fallback method (using textured quads).

I noticed that due to Issue #235, the fire texture is loaded lazily. But this does not offer the opportunity to clients code to set their own texture. Maybe it would be best to have the texture class attribute contain the file name for the texture, and load it during the `__init__` method. Something along the line of

```python
def load_texture(self):
    pic = pyglet.image.load(self.texture, file=pyglet.resource.file(self.texture))
    self.__class__.texture = pic.get_texture()
```
Daniel.

**Comment by ccanepa** [comment]
Can you commit to master the shader changes and close this issue ?
Then make another issue about custom textures ?

**Comment by dangillet** [comment]
Issue fixed in commit 8804326e

**Comment by dangillet** [comment]
Issue fixed in commit 8804326e
